### PR TITLE
DNM os/bluestore: bluefs write cache

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -1919,9 +1959,9 @@ int BlueFS::open_for_write(
     // match up with bluestore.  the slow device is always the second
     // one (when a dedicated block.db device is present and used at
     // bdev 0).  the wal device is always last.
-    if (boost::algorithm::ends_with(filename, ".slow")) {
+    if (strcmp(dirname.c_str() + dirname.length() - 5, ".slow") == 0) {
```
Comment (varadakari) [inline src/os/bluestore/BlueFS.cc]: can we use dirname.data() instead of c_str()

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -1919,9 +1959,9 @@ int BlueFS::open_for_write(
...
       file->fnode.prefer_bdev = BlueFS::BDEV_SLOW;
-    } else if (boost::algorithm::ends_with(dirname, ".wal")) {
+    } else if (strcmp(dirname.c_str() + dirname.length() - 4, ".wal") == 0) {
```
Comment (varadakari) [inline src/os/bluestore/BlueFS.cc]: same here

File: src/os/bluestore/BlueFS.cc [hunk 3]
```diff
@@ -2179,7 +2264,7 @@ int BlueFS::readdir(const string& dirname, vector<string> *ls)
 {
   std::lock_guard<std::mutex> l(lock);
   dout(10) << __func__ << " " << dirname << dendl;
-  if (dirname.empty()) {
+  if (dirname.size() == 0) {
```
Comment (varadakari) [inline src/os/bluestore/BlueFS.cc]: empty() or size() both are constant complexity. Anything specific in using size()?

File: src/os/bluestore/BlueFS.cc [hunk 4]
```diff
@@ -1971,9 +2013,54 @@ void BlueFS::_close_writer(FileWriter *h)
...
+    _write_cache_add(h->file);
+  }
+
   delete h;
 }
 
+void BlueFS::_write_cache_add(FileRef file)
+{
+  // add to write cache
+  write_cache_bytes += file->contents.length();
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: Shouldn't we bypass adding to the cache if bluefs_write_cache_bytes == 0 ( or < file->contents.length())?

File: src/os/bluestore/BlueFS.cc [hunk 5]
```diff
@@ -901,6 +921,19 @@ int BlueFS::_read(
            << " 0x" << std::hex << off << "~" << len << std::dec
 	   << " from " << h->file->fnode << dendl;
 
+  logger->inc(l_bluefs_reads);
+  logger->inc(l_bluefs_read_bytes, len);
+  if (h->file->contents.length()) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: Wondering if it's possible to have less data in file->contents than off+len? 
As far as I understand we are adding just portions of content to the cache in _flush_range so the question is if there might be a _read call for range(s) not handled by _flush_range yet.

Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: OK, the latter isn't possible due to assert(h->file->contents.length() == offset + length); in _flush_range

**Comment by liewegas** [comment]
In C++11/C++0x, data() and c_str() are no longer different

**Comment by ifed01** [comment]
Fails for me at

**Comment by ifed01** [comment]
another one:
ObjectStore/StoreTestSpecificAUSize.SyntheticMatrixNoCsum/2
...
---------------------- 11 / 16 ----------------------
  bluestore_min_alloc_size = 65536
  max_write = 65536
  max_size = 1048576
  alignment = 512
  bluestore_max_blob_size = 262144
  bluestore_compression_mode = force
  bluestore_csum_type = none
  bluestore_default_buffered_read = false
  bluestore_default_buffered_write = true
  bluestore_sync_submit_transaction = true
seeding object 0
seeding object 500
Op 0
available_objects: 998 in_flight_objects: 2 total objects: 1000 in_flight 2
Op 1000
available_objects: 987 in_flight_objects: 5 total objects: 992 in_flight 5
Op 2000
available_objects: 995 in_flight_objects: 1 total objects: 996 in_flight 1
Op 3000
available_objects: 1002 in_flight_objects: 0 total objects: 1002 in_flight 0
Op 4000
available_objects: 1012 in_flight_objects: 0 total objects: 1012 in_flight 0
2017-04-04 12:25:56.848303 7f4965b98a80 -1 bluestore(bluestore.test_temp_dir) fsck shared blob 0x6f9 ref_map ref_map(0x417a0000~10000=3) != expected ref_map(0x417a0000~10000=2)
2017-04-04 12:25:56.848329 7f4965b98a80 -1 bluestore(bluestore.test_temp_dir) fsck shared blob 0x6fa ref_map ref_map(0x41a60000~10000=3) != expected ref_map(0x41a60000~10000=2)
2017-04-04 12:25:56.849198 7f4965b98a80 -1 bluestore(bluestore.test_temp_dir) fsck actual store_statfs(0x21c010000/0x2625a0000, stored 0x1e8302a/0x4c90000, compress 0x0/0x0/0x0) != expected store_statfs(0x21c010000/0x2625a0000, stored 0x1e5be2a/0x4c40000, compress 0x0/0x0/0x0)
2017-04-04 12:25:56.904148 7f4965b98a80 -1 bluestore(bluestore.test_temp_dir) fsck leaked some space;5242880 bytes leaked
ceph_test_objectstore: /root/ceph/ceph2/src/test/objectstore/store_test.cc:4013: void SyntheticWorkloadState::fsck(bool): Assertion `r == 0 || r == -EOPNOTSUPP' failed.
*** Caught signal (Aborted) **

**Comment by liewegas** [comment]
http://pulpito.ceph.com/yuriw-2017-04-17_21:11:42-rados-wip-sage-testing2_2017_4_18-distro-basic-smithi/1036915

<pre>
2017-04-18T01:42:29.585 INFO:teuthology.orchestra.run.smithi160.stderr:ceph-objectstore-tool: /build/ceph-12.0.0-2684-ge82c844/obj-x86_64-linux-gnu/boost/include/boost/intrusive/detail/generic_hook.hpp:48: void boost::intrusive::detail::destructor_impl(Hook&, boost::intrusive::detail::link_dispatch<(boost::intrusiv
e::link_mode_type)1u>) [with Hook = boost::intrusive::generic_hook<(boost::intrusive::algo_types)0u, boost::intrusive::list_node_traits<void*>, boost::intrusive::member_tag, (boost::intrusive::link_mode_type)1u, (boost::intrusive::base_hook_type)0u>]: Assertion `!hook.is_linked()' failed.
2017-04-18T01:42:29.586 INFO:teuthology.orchestra.run.smithi160.stderr:*** Caught signal (Aborted) **
2017-04-18T01:42:29.586 INFO:teuthology.orchestra.run.smithi160.stderr: in thread 7f76d9ea5f40 thread_name:ceph-objectstor
2017-04-18T01:42:29.596 INFO:teuthology.orchestra.run.smithi160.stderr: ceph version 12.0.0-2684-ge82c844 (e82c84421a5093fef5afc6a09ba4cc34a825adf6)
2017-04-18T01:42:29.596 INFO:teuthology.orchestra.run.smithi160.stderr: 1: (()+0xa956f7) [0x7f76d96cf6f7]
2017-04-18T01:42:29.596 INFO:teuthology.orchestra.run.smithi160.stderr: 2: (()+0x10330) [0x7f76ceed6330]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 3: (gsignal()+0x37) [0x7f76ce10fc37]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 4: (abort()+0x148) [0x7f76ce113028]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 5: (()+0x2fbf6) [0x7f76ce108bf6]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 6: (()+0x2fca2) [0x7f76ce108ca2]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 7: (()+0xa44190) [0x7f76d967e190]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 8: (BlueFS::File::~File()+0x70) [0x7f76d969c3a0]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 9: (RefCountedObject::put() const+0xe9) [0x7f76d91b52d9]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 10: (BlueFS::unlink(std::string const&, std::string const&)+0x196) [0x7f76d9699626]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 11: (BlueRocksEnv::DeleteFile(std::string const&)+0x104) [0x7f76d96a6424]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 12: (rocksdb::DeleteSSTFile(rocksdb::ImmutableDBOptions const*, std::string const&, unsigned int)+0x80) [0x7f76d988e530]
2017-04-18T01:42:29.597 INFO:teuthology.orchestra.run.smithi160.stderr: 13: (rocksdb::DBImpl::DeleteObsoleteFileImpl(rocksdb::Status, int, std::string const&, rocksdb::FileType, unsigned long, unsigned int)+0x46) [0x7f76d96f57f0]
2017-04-18T01:42:29.598 INFO:teuthology.orchestra.run.smithi160.stderr: 14: (rocksdb::DBImpl::PurgeObsoleteFiles(rocksdb::JobContext const&, bool)+0xd54) [0x7f76d96f6828]
2017-04-18T01:42:29.598 INFO:teuthology.orchestra.run.smithi160.stderr: 15: (rocksdb::DBImpl::~DBImpl()+0x3e3) [0x7f76d96f3425]
2017-04-18T01:42:29.598 INFO:teuthology.orchestra.run.smithi160.stderr: 16: (rocksdb::DBImpl::~DBImpl()+0x18) [0x7f76d96f3a18]
2017-04-18T01:42:29.598 INFO:teuthology.orchestra.run.smithi160.stderr: 17: (RocksDBStore::~RocksDBStore()+0x4e) [0x7f76d963e36e]
2017-04-18T01:42:29.598 INFO:teuthology.orchestra.run.smithi160.stderr: 18: (RocksDBStore::~RocksDBStore()+0x9) [0x7f76d963e659]
</pre>

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
<pre>
2017-04-21T02:52:15.529 INFO:tasks.ceph.osd.4.smithi003.stderr:2017-04-21 02:52:15.559117 7f19b8247700 -1 rocksdb: submit_transaction error: Corruption: block checksum mismatch code = 2 Rocksdb transaction:
2017-04-21T02:52:15.529 INFO:tasks.ceph.osd.4.smithi003.stderr:Delete( Prefix = O key = 0x7f7ffffffffffffffe48000000217363'rub_0.12!='0xfffffffffffffffeffffffffffffffff'o')
2017-04-21T02:52:15.530 INFO:tasks.ceph.osd.4.smithi003.stderr:/mnt/jenkins/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.1-1260-g83f52c7/rpm/el7/BUILD/ceph-12.0.1-1260-g83f52c7/src/os/bluestore/BlueStore.cc: In function 'void BlueS
tore::_kv_sync_thread()' thread 7f19b8247700 time 2017-04-21 02:52:15.559134
2017-04-21T02:52:15.530 INFO:tasks.ceph.osd.4.smithi003.stderr:/mnt/jenkins/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.1-1260-g83f52c7/rpm/el7/BUILD/ceph-12.0.1-1260-g83f52c7/src/os/bluestore/BlueStore.cc: 7899: FAILED assert(r =
= 0)
2017-04-21T02:52:15.530 INFO:tasks.ceph.osd.4.smithi003.stderr: ceph version 12.0.1-1260-g83f52c7 (83f52c744d61badf26432f57463916e6d2e86b3c)
2017-04-21T02:52:15.530 INFO:tasks.ceph.osd.4.smithi003.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x110) [0x7f19cae0fc80]
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr: 2: (BlueStore::_kv_sync_thread()+0x2bfe) [0x7f19cacb131e]
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr: 3: (BlueStore::KVSyncThread::entry()+0xd) [0x7f19cace84bd]
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr: 4: (()+0x7dc5) [0x7f19c7a5cdc5]
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr: 5: (clone()+0x6d) [0x7f19c6b5073d]
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2017-04-21T02:52:15.531 INFO:tasks.ceph.osd.4.smithi003.stderr:2017-04-21 02:52:15.560841 7f19b8247700 -1 /mnt/jenkins/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.1-1260-g83f52c7/rpm/el7/BUILD/ceph-12.0.1-1260-g83f52c7/src/os/blue
</pre>

**Pull request closed without merging**